cli/command/trust/signer_add.go

```diff
@@ -0,0 +1,138 @@
+	flags.VarP(&options.keys, "key", "k", "Path to the signer's public key(s)")
+	return cmd
+}
+
+var validSignerName = regexp.MustCompile(`^[a-z0-9]+[a-z0-9\_\-]*$`).MatchString
+
+func addSigner(cli command.Cli, options signerAddOptions) error {
+	signerName := options.signer
+	if !validSignerName(signerName) {
+		return fmt.Errorf("signer name \"%s\" must not contain uppercase or special characters", signerName)
+	}
+	if signerName == "releases" {
+		return fmt.Errorf("releases is a reserved keyword, please use a different signer name")
+	}
+
+	if options.keys.Len() < 1 {
+		return fmt.Errorf("path to a valid public key must be provided using the `--key` flag")
+	}
+
+	var errImages []string
+	for _, imageName := range options.images {
+		if err := addSignerToImage(cli, signerName, imageName, options.keys.GetAll()); err != nil {
+			fmt.Fprintln(cli.Err(), err.Error())
+			errImages = append(errImages, imageName)
+		} else {
+			fmt.Fprintf(cli.Out(), "Successfully added signer: %s to %s\n", signerName, imageName)
+		}
+	}
+	if len(errImages) > 0 {
+		return fmt.Errorf("Failed to add signer to: %s", strings.Join(errImages, ", "))
+	}
+	return nil
+}
+
+func addSignerToImage(cli command.Cli, signerName string, imageName string, keyPaths []string) error {
+	fmt.Fprintf(cli.Out(), "\nAdding signer \"%s\" to %s...\n", signerName, imageName)
+
+	ctx := context.Background()
+	imgRefAndAuth, err := trust.GetImageReferencesAndAuth(ctx, image.AuthResolver(cli), imageName)
+	if err != nil {
+		return err
+	}
+
+	notaryRepo, err := cli.NotaryClient(imgRefAndAuth, trust.ActionsPushAndPull)
+	if err != nil {
+		return trust.NotaryError(imgRefAndAuth.Reference().Name(), err)
+	}
+
+	if _, err = notaryRepo.ListTargets(); err != nil {
+		switch err.(type) {
+		case client.ErrRepoNotInitialized, client.ErrRepositoryNotExist:
+			fmt.Fprintf(cli.Out(), "Initializing signed repository for %s...\n", imageName)
+			if err := getOrGenerateRootKeyAndInitRepo(notaryRepo); err != nil {
+				return trust.NotaryError(imageName, err)
+			}
+			fmt.Fprintf(cli.Out(), "Successfully initialized %q\n", imageName)
+		default:
+			return trust.NotaryError(imageName, err)
+		}
+	}
+
+	newSignerRoleName := data.RoleName(path.Join(data.CanonicalTargetsRole.String(), signerName))
+
+	signerPubKeys, err := ingestPublicKeys(keyPaths)
+	if err != nil {
+		return err
+	}
+	if err := addStagedSigner(notaryRepo, newSignerRoleName, signerPubKeys); err != nil {
+		return errors.Wrapf(err, "could not add signer to repo: %s", strings.TrimPrefix(newSignerRoleName.String(), "targets/"))
+	}
+
+	return notaryRepo.Publish()
+}
+
+func ingestPublicKeys(pubKeyPaths []string) ([]data.PublicKey, error) {
+	pubKeys := []data.PublicKey{}
+	for _, pubKeyPath := range pubKeyPaths {
+		// Read public key bytes from PEM file
+		pubKeyBytes, err := ioutil.ReadFile(pubKeyPath)
+		if err != nil {
+			if os.IsNotExist(err) {
+				return nil, fmt.Errorf("file for public key does not exist: %s", pubKeyPath)
```

Comment:
Looks like we don't have to special case `not exist` errors. Standard errors look to provide the right information. When removing the special case, it looks like:

```
unable to read public key from file: /no-such-key: open /no-such-key: no such file or directory

unable to read public key from file: /foo/foobar.pub: open /foo/foobar.pub: permission denied
```

So no `Wrapf()` needed as well, just:

```go
return nil, errors.Wrap(err, "unable to read public key from file")
```